### PR TITLE
Add analytics support for password grant type

### PR DESCRIPTION
**Pull request by ShanChathusanda93**
### Proposed changes in this pull request

* Fix : https://github.com/wso2/product-is/issues/5065
* Added authentication publish methods for password grant type to publish authentication events to analytics.
* Changed a test case to support the added publish methods.

### When should this PR be merged

* Please merge this PR after https://github.com/wso2-extensions/identity-data-publisher-authentication/pull/81

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java [hunk 1]
```diff
@@ -174,6 +200,86 @@ private boolean validateUserCredentials(OAuth2AccessTokenReqDTO tokenReq, Servic
...
+        String contextId = UUIDGenerator.generateUUID();
+        authenticationContext.setContextIdentifier(contextId);
+        authenticationContext.setTenantDomain(authenticatedUser.getTenantDomain());
+        authenticationContext.setRequestType(OAUTH2);
+        authenticationContext.setRememberMe(false);
+        authenticationContext.setForceAuthenticate(true);
+        authenticationContext.setPassiveAuthenticate(false);
+        authenticationContext.setProperty(IS_INITIAL_LOGIN, true);
+
+        // setting sequenceConfig with authenticatedUser, serviceProvider
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java:254]: Start with a capital letter and end with a full stop for the comments. Do this in all applicable places.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java [hunk 2]
```diff
@@ -174,6 +200,86 @@ private boolean validateUserCredentials(OAuth2AccessTokenReqDTO tokenReq, Servic
...
+
+        // setting applicationConfig with serviceProvider
+        ApplicationConfig applicationConfig = new ApplicationConfig(serviceProvider);
+        sequenceConfig.setApplicationConfig(applicationConfig);
+
+        sequenceConfig.setAuthenticatedIdPs(FrameworkConstants.LOCAL_IDP_NAME);
+        authenticationContext.setSequenceConfig(sequenceConfig);
+
+        // setting the authenticated IDP for currentAuthenticatedIDPs to get
+        // the tenant domain and other parameters when the login is a success
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java:266]: Use java multiline comment

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 3]
```diff
@@ -391,4 +393,25 @@ protected void unsetTokenBinderInfo(TokenBinderInfo tokenBinderInfo) {
...
     }
+
+    @Reference(
+            name = "framework.authentication.data.publisher",
+            service = AuthenticationDataPublisher.class,
+            cardinality = ReferenceCardinality.MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetAuthenticationDataPublisher"
+    )
+    protected void setAuthenticationDataPublisher(AuthenticationDataPublisher dataPublisher) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:404]: Add a new line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 4]
```diff
@@ -391,4 +393,25 @@ protected void unsetTokenBinderInfo(TokenBinderInfo tokenBinderInfo) {
...
+        if (FrameworkConstants.AnalyticsAttributes.AUTHN_DATA_PUBLISHER_PROXY.equalsIgnoreCase(dataPublisher.
+                getName()) && dataPublisher.isEnabled(null)) {
+            OAuth2ServiceComponentHolder.setAuthenticationDataPublisherProxy(dataPublisher);
+        }
+    }
+
+    protected void unsetAuthenticationDataPublisher(AuthenticationDataPublisher dataPublisher) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:411]: Add a new line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 5]
```diff
@@ -391,4 +393,25 @@ protected void unsetTokenBinderInfo(TokenBinderInfo tokenBinderInfo) {
...
+    @Reference(
+            name = "framework.authentication.data.publisher",
+            service = AuthenticationDataPublisher.class,
+            cardinality = ReferenceCardinality.MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetAuthenticationDataPublisher"
+    )
+    protected void setAuthenticationDataPublisher(AuthenticationDataPublisher dataPublisher) {
+        if (FrameworkConstants.AnalyticsAttributes.AUTHN_DATA_PUBLISHER_PROXY.equalsIgnoreCase(dataPublisher.
+                getName()) && dataPublisher.isEnabled(null)) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:406]: Why do we pass null to the isEnabled() method?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java [hunk 6]
```diff
@@ -198,4 +200,23 @@ public void setOAuthAdminService(OAuthAdminServiceImpl oauthAdminService) {
...
         this.oauthAdminService = oauthAdminService;
     }
+
+    /**
+     * Set Authentication Data Publisher Proxy instance.
+     *
+     * @param authenticationDataPublisherProxy
+     */
+    public static void setAuthenticationDataPublisherProxy(AuthenticationDataPublisher
+                                                                   authenticationDataPublisherProxy) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java:210]: Add an empty line as the 1st line of the method. Check in other places as well.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java (hunk 1 above)
Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java]: Modified

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java (hunk 2 above)
Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java]: Modified.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java (hunk 3 above)
Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:404]: Modified.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java (hunk 4 above)
Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:412]: Modified.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java (hunk 6 above)
Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java:210]: Modified.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java [hunk 7]
```diff
@@ -174,6 +200,86 @@ private boolean validateUserCredentials(OAuth2AccessTokenReqDTO tokenReq, Servic
...
     }
 
+    /**
+     * This method will publish the Password Grant Authentication data.
+     *
+     * @param tokenReq Token request which contains all the details of the request.
+     * @param authenticated Boolean value which determines whether the user is authenticated or not.
+     * @param serviceProvider Service provider which contains the details of the application.
+     */
+    private void publishAuthenticationData(OAuth2AccessTokenReqDTO tokenReq, boolean authenticated,
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java:210]: Shall we make this method protected?
Will be useful for anyone extending the grant handler.

Comment (ShanChathusanda93) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java]: Modified.

**Pull request merged**